[android] turn on USE_VULKAN for android builds by default

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51291 [android] turn on USE_VULKAN for android builds by default**

Turning on USE_VULKAN for android builds
Remove standalone android vulkan build

Testing all ci jobs (for master): https://github.com/pytorch/pytorch/pull/51292

Differential Revision: [D26141891](https://our.internmc.facebook.com/intern/diff/D26141891)

IvanKobzarev committed:
[android] turn on USE_VULKAN for android builds by default

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] turn on USE_VULKAN for android builds by default"




[ghstack-poisoned]

AshkanAliabadi commented on CMakeLists.txt:
```diff
@@ -219,7 +219,9 @@ option(USE_TENSORRT "Using Nvidia TensorRT library" OFF)
-option(USE_VULKAN "Use Vulkan GPU backend" OFF)
+cmake_dependent_option(
+    USE_VULKAN "Use Vulkan GPU backend" ON
+    "ANDROID" OFF)
```
Is USE_VULKAN dependent on ANDROID being OFF or ON?  Will this be OK for desktop builds?

IvanKobzarev replied:
https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html
Here the argument positions are non trivial: the default value is the latest, ON if ANDROID

AshkanAliabadi replied:
Got your reply in an email, so is my understanding correct that this statement means: give the users the option to set USE_VULKAN to ON or OFF if ANDROID is OFF, right?

IvanKobzarev replied:
Yes, I deleted my reply to verify if overrides with -DUSE_VULKAN work correct.

And in that case it does not. It is ON if android and OFF else.

I will try to find the way how to make override work

IvanKobzarev replied:
@AshkanAliabadi , updated to construction
```
if(NOT DEFINED USE_VULKAN)
  cmake_dependent_option(
      USE_VULKAN "Use Vulkan GPU backend" ON
      "ANDROID" OFF)
endif()
```
which works fine with -DUSE_VULKAN=

IvanKobzarev committed:
Update on "[android] turn on USE_VULKAN for android builds by default"


Turning on USE_VULKAN for android builds
Remove standalone android vulkan build

Testing all ci jobs (for master): https://github.com/pytorch/pytorch/pull/51292


[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] turn on USE_VULKAN for android builds by default"


Turning on USE_VULKAN for android builds
Remove standalone android vulkan build

Testing all ci jobs (for master): https://github.com/pytorch/pytorch/pull/51292

Differential Revision: [D26141891](https://our.internmc.facebook.com/intern/diff/D26141891)

[ghstack-poisoned]

AshkanAliabadi commented on CMakeLists.txt:
```diff
@@ -219,7 +219,13 @@ option(USE_TENSORRT "Using Nvidia TensorRT library" OFF)
+
 option(USE_VULKAN_FP16_INFERENCE "Vulkan - Use fp16 inference even on fp32 tensors" OFF)
 option(USE_VULKAN_RELAXED_PRECISION "Vulkan - Use relaxed precision math in the kernels (mediump)" OFF)
 option(USE_VULKAN_SHADERC_RUNTIME "Vulkan - Use runtime shader compilation (needs libshaderc)" OFF)
```
I think it will also makes sense to make these series of flags dependent on USE_VULKAN being ON.  What do you think

IvanKobzarev replied:
I think for the first iteration we can keep it like we have - USE_VULKAN works as a master switch anyway.

AshkanAliabadi reviewed (approved):
Thanks Ivan!

IvanKobzarev committed:
Update on "[android] turn on USE_VULKAN for android builds by default"


Turning on USE_VULKAN for android builds
Remove standalone android vulkan build

Testing all ci jobs (for master): https://github.com/pytorch/pytorch/pull/51292

Differential Revision: [D26141891](https://our.internmc.facebook.com/intern/diff/D26141891)

[ghstack-poisoned]
